File: binary_search/almost_duplicates/almost_duplicates.cpp
#include <vector>
#include <queue>
using namespace std;

bool count_almost_duplicates(vector<int> &nums, int idx_diff, int val_diff)
{
    // LeetCode Q.220.
    queue<int> queue;   // Queue of indices of nums in current window.
    vector<int> window; // Sorted nums within current window.

    for (int idx = 0; idx < nums.size(); idx++)
    {
        if (!queue.empty() && queue.front() < idx - idx_diff)
        {
            int window_idx = upper_bound(window.begin(), window.end(), nums[queue.front()]) - window.begin();
            window.erase(window.begin() + window_idx - 1);
            queue.pop();
        }

        int window_idx = upper_bound(window.begin(), window.end(), nums[idx]) - window.begin();

        if (window_idx < window.size())
        {
            if (abs(window[window_idx] - nums[idx]) <= val_diff)
            {
                return true;
            }
        }

        if (0 < window_idx)
        {
            if (abs(window[window_idx - 1] - nums[idx]) <= val_diff)
            {
                return true;
            }
        }

        window.insert(window.begin() + window_idx, nums[idx]);
        queue.push(idx);
    }

    return false;
}

File: binary_search/almost_duplicates/almost_duplicates.py

def _binary_search(target: int, sorted_integers: list[int], size: int) -> int:
    left_idx, right_idx = 0, size - 1
    while left_idx <= right_idx:
        mid_idx = (left_idx + right_idx) // 2
        if sorted_integers[mid_idx] < target:
            left_idx = mid_idx + 1
            continue
        right_idx = mid_idx - 1

    return left_idx  # Number of ints < target, implying insertion idx.


def count_almost_duplicates(integers: list[int], idx_diff: int, val_diff: int) -> bool:  # LeetCode Q.220.
    """Find if any pair of ints within idx diff and value diff."""
    total_nums = len(integers)
    window_size = min(total_nums, idx_diff + 1)
    window_ints = integers[: window_size]  # Window of ints within idx difference.
    window_ints.sort()  # Sort for binary search.

    for i in range(window_size - 1):
        if window_ints[i + 1] - window_ints[i] <= val_diff:
            return True

    current_idx = 1
    while current_idx + idx_diff < total_nums:
        pop_idx = _binary_search(integers[current_idx - 1], window_ints, window_size)
        window_ints.pop(pop_idx)  # Window's 1st int is out.
        window_size -= 1

        newcomer = integers[current_idx + idx_diff]  # The int that is joining window.
        insertion_idx = _binary_search(newcomer, window_ints, window_size)

        if insertion_idx == 0:  # Newcomer < all window ints.
            if window_ints[0] - newcomer <= val_diff:
                return True

        if insertion_idx == window_size:  # Newcomer > all window ints.
            if newcomer - window_ints[-1] <= val_diff:
                return True

        if 0 < insertion_idx < window_size:
            if min(newcomer - window_ints[insertion_idx - 1], window_ints[insertion_idx] - newcomer) <= val_diff:
                return True

        window_ints.insert(insertion_idx, newcomer)
        window_size += 1
        current_idx += 1

    return False
